docs/design/auditing-containment.md

```diff
@@ -0,0 +1,205 @@
+
+For 1, the node doesn't necessarily need to be contained. For 2, the node does need to be contained.
+For 3 and 4, the node should probably also be contained.
+
+We need to find out what the error messages are for these different cases so that we don't lump the first case with the rest.
+
+### Handling Contained Nodes
+New functionality needs to be added to the audit verifier, where upon receiving a pointer, the verifier should first iterate through the nodes listed in the pointer and check if they have a `contained` field set to true on their NodeDossier or not.
+
+For nodes that are not contained, the audit will continue as normal.
+
+For nodes that are contained, the verifier will ask the ContainmentDB for the hash of the erasure share expected from the node when it was originally audited.
+
+The verifier will then attempt to download that erasure share from the contained node.
+From there, the following cases can occur:
+- Passes audit ->  `pendingAudit` is deleted from the ContainmentDB  -> stats updated in overlaycache/satelliteDB
+- Fails audit -> `pendingAudit` is deleted from the ContainmentDB -> stats updated in overlaycache/satelliteDB
+- Refuses to respond again -> the `pendingAudit`'s `reverifyCount` is incremented in the ContainmentDB ->
+    - -> If that count exceeds the reverify limit, then the node’s stats will be updated to reflect an audit failure and the node will be removed from the ContainmentDB.
+    - -> If the count does not exceed the reverify limit, the `pendingAudit` will remain in the ContainmentDB.
+
+```go
+type pendingAudit struct {
+    nodeID            storj.NodeID
+    pieceID           storj.PieceID
+    pieceNum          int
+    stripeIndex       int
+    shareSize         int64
+    expectedShareHash []byte
+    reverifyCount     int // number of times reverify has been attempted
+}
+```
+`pendingAudit` entries are created and added to the ContainmentDB when a node seems to be online and opens a connection with the satellite to be audited (this happens within the existing Verify function in the audit package), but then refuses to send the requested erasure share.
+
+Additionally, the satellite should make sure to empty the ContainmentDB of certain pending audits when segments are deleted.
+
+```go
+type ContainmentDB interface {
+  Get(ctx context.Context, nodeID pb.NodeID) (pendingAudit, error)
+  IncrementPending(ctx context.Context, pendingAudit pendingAudit) error
+  Delete(ctx context.Context, nodeID pb.NodeID) error
+}
+```
+
+`Get` will be used in the Reverify function to get the `pendingAudit` information from the ContainmentDB needed to compare the expected audit result with the actual.
+
+`IncrementPending` is an upsert that is used when an "uncooperative" node is identified during the normal process, and when a node doesn't pass a reverification.
+
+`Delete` will be used to delete the pendingAudit from the ContainmentDB after the node either passes or fails the audit.
```

Comment:
Maybe it could also update the node dossier

Reply:
that's a good idea!